src/bitcoin/electrum/wallet.rs

```diff
@@ -0,0 +1,458 @@
+                    .graph
+                    .graph()
+                    .get_chain_position(&self.local_chain, self.local_chain.tip().block_id(), txid)
+                {
+                    last_seen
+                } else {
+                    last_seen
+                };
+                log::debug!(
+                    "changing last seen for txid '{}' from {} to {}",
+                    txid,
+                    prev_last_seen,
+                    now
+                );
+                let _ = graph_update.insert_seen_at(txid, now);
+            }
+        }
+        let _ = self.graph.apply_update(graph_update);
+        Ok(())
+    }
+
+    /// Get the block that `tip` has in common with the Electrum server
+    /// based on the wallet's checkpoints.
+    pub fn common_ancestor_with_electrum(
+        &self,
+        client: &Client,
+        tip: &BlockChainTip,
+    ) -> Option<BlockChainTip> {
+        let server_tip_height = chain_tip(client).height as u32;
+        let tip_block = BlockId {
+            hash: tip.hash,
+            height: height_u32_from_i32(tip.height),
+        };
+        // Get a local chain that includes all our checkpoints up to and including `tip`.
+        // Typically, the local chain's tip should be the same as `tip`, but the local chain's tip
+        // may have advanced slightly in case the Electrum tip changed while performing a round of
+        // updates and we restarted.
+        let local_chain = {
+            let mut chain = self.local_chain.clone();
+            // We want to disconnect all checkpoints *after* `tip`, but `disconnect_from` is inclusive.
+            // So call `disconnect_from` and then re-insert `tip`.
+
+            // We can only get an error if `tip` is the genesis block, but we should never
+            // be trying to find the common ancestor of the genesis block.
+            let _ = chain
+                .disconnect_from(tip_block)
+                .expect("we should not be trying to find common ancestor with genesis block");
+            let _ = chain
+                .insert_block(tip_block)
+                .expect("we have already removed this block from chain");
+            chain
+        };
+        // The following code is based on the function `construct_update_tip`. See:
+        // https://github.com/bitcoindevkit/bdk/blob/4a8452f9b8f8128affbb60665016fedb48f07cd6/crates/electrum/src/electrum_ext.rs#L284
+        // TODO: Is the following comment and code correct in our case? Could the electrum tip be lower and a reorged chain?
+
+        // If electrum returns a tip height that is lower than our previous tip, then checkpoints do
+        // not need updating. We just return the previous tip and use that as the point of agreement.
+        // if new_tip_height < prev_tip.height() {
+        //     return Ok((prev_tip.clone(), Some(prev_tip.height())));
+        // }
```

Comment:
I commented out this logic for now as I wasn't sure if it was always true in case of a reorg. But this logic is still part of the BDK `sync()` method.

Reply:
No longer relevant: we removed the common ancestor logic for Electrum.